### ISSUE TITLE
Strange AHRS results

### ISSUE THREAD

**Issue by nitrosalat** [kriswiner/MPU9250#220]
Hi, i'm sorry if this question is asked before.

When mpu9250 module accelerates along x-axis and not rotates around y-axes pitch value shouldn't change. But in my case pitch value changes. What am I doing wrong?

**Comment by kriswiner** [comment]
Did you calibrate all thre sensors?

**Comment by nitrosalat** [comment]
Yes, i did

**Comment by kriswiner** [comment]
Are you passing sensor data to the Madgwick filter in the proper order?

**Comment by nitrosalat** [comment]
Yes, i tried Madgwick and Mahony filters and i'm passing sensor data in proper order. The acceleration still affects the Euler angles.

**Comment by kriswiner** [comment]
What is the order?

If it correct, and I assume your sensors are properly calbrated, only other
possiblity apart from a faulty sensor breakout would be man-made magnetic
fields.

**Comment by nitrosalat** [comment]
ax, ay, az, gx, gy, gz, mx, my, mz, deltat

**Comment by kriswiner** [comment]
No, needs to be NED.

**Comment by nitrosalat** [comment]
In static position(without accelerations) pitch, roll angles are correct but if i accelerate the module along one of axes WITHOUT rotation pitch and roll are changing. I think it is not a problem with sensor data pass order to filter.

**Comment by kriswiner** [comment]
" I think it is not a problem with sensor data pass order to filter."

You are wrong.

Choose which accel axis you want to point North. Figure out which gyro and
mag axes point in the same direction. Then figgure out which oint East, and
which point Down. If you look at the data sheet you should be able to see
immediately why this is necessary; i..e.,Az = -Mz in terms of direction.

Feed the Madgwick filter as so:

(AN, AE, AD, GN, GE, GD, MN, ME, MD)

It cannot be as you had it above.

**Comment by nitrosalat** [comment]
Ok, i'm feeding the madgwick filter in order ax, ay, az, gx, gy, gz, my, mx, -mz but pitch and roll values are still changing while the module accelerates along x-axis or along y-axis. 
Btw, when i use madgwick filter, pitch and roll less react on acceleration than mahony filter.

**Comment by kriswiner** [comment]
What kind of MCU are you using? What is the fusion rate?

**Comment by nitrosalat** [comment]
ATmega328P(arduino nano)

**Comment by kriswiner** [comment]
Yeah, 8 or 16 MHz AVR is not going to give you very good results. You might
get 100 Hz fusion rates, you really need 1 kHz or better.

See:
https://github.com/kriswiner/MPU6050/wiki/Affordable-9-DoF-Sensor-Fusion

Teensy 3.2 or, better, this
<https://www.tindie.com/products/TleraCorp/ladybug-stm32l432-development-board/>
would be more appropriate.

**Comment by nitrosalat** [comment]
Thanks, Kris :)

**Comment by kriswiner** [comment]
Sorry there is no quick fix.

Accurate absolute orientation is hard and there are no short cuts,
unfortunately ;<
